### PR TITLE
backupccl: consider followers in backup proc planning

### LINKED ISSUE #114528: backupccl: follower reads not considered during distsql planning
During DistSQL planning the `followerReadOracle` is only picked under some conditions outlined over here - https://github.com/cockroachdb/cockroach/blob/cf5c2caa84c635acbc22e73c8f2db35aa9de6c71/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go#L196-L196. One of the conditions is `txn != nil`. Both backup and changefeed do not plan their DistSQL flows in the scope of a txn, and so pass in nil txns to `dsp. SetupAllNodesPlanningWithOracle`. This means that we always pick the `binPackingOracle` instead that prefers leaseholders - https://github.com/cockroachdb/cockroach/blob/3111a55e91dba9a88931fda725b767163c981ae7/pkg/sql/physicalplan/replicaoracle/oracle.go#L270.

When a backup is run with execution locality it constrains the execution and planning of the backup to only the nodes in the specified region. If the user has configured the backup to run in a non-primary region with follower replicas, we should be picking these replicas instead of reaching out to the leaseholders in the primary region. This way users can avoid backups interfering with foreground traffic. This is not likely today because of our inability to use the `followerReadOracle`.

The likely fix for this bug is to plumb a timestamp or txn from backup, cdc flow planning to where we can decide to use follower reads or not. 

Jira issue: CRDB-33547

### PR DESCRIPTION
**Pull request by dt**
Followers are only considered when planning procesor placement if the txn passed to the replica oracle is non-nil, as the txn timestamp is what is used to determine if it is able to be served by a follower.

This changes backup to pass a non-nil txn that is used exclusively to carry this timestamp from the backup process to the planning oracle.

Release note (bug fix): backups will now plan where to execute taking follower replicas into account.
Epic: none.

### CODE REVIEW

File: pkg/sql/distsql_plan_bulk.go [hunk 1]
```diff
@@ -46,9 +47,26 @@ func (dsp *DistSQLPlanner) SetupAllNodesPlanningWithOracle(
...
+// SetupAllNodesPlanningWithOracleAndFollowers is like
+// SetupAllNodesPlanningWithOracle but with a txn param that if non-nil allows
+// the flow to plan on followers.
+func (dsp *DistSQLPlanner) SetupAllNodesPlanningWithOracleAndFollowers(
+	ctx context.Context,
+	evalCtx *extendedEvalContext,
+	execCfg *ExecutorConfig,
+	oracle replicaoracle.Oracle,
+	localityFilter roachpb.Locality,
+	txn *kv.Txn,
```
Comment (miretskiy) [inline pkg/sql/distsql_plan_bulk.go:64]: Is there a way to pass in a timestamp? Or, if not (because of backport considerations), at least leave a TODO to not rely on txn?

File: pkg/sql/distsql_plan_bulk.go [hunk 2]
```diff
@@ -46,9 +47,26 @@ func (dsp *DistSQLPlanner) SetupAllNodesPlanningWithOracle(
...
+		return dsp.setupAllNodesPlanningSystem(ctx, evalCtx, execCfg, oracle, localityFilter, nil /*txn*/)
 	}
-	return dsp.setupAllNodesPlanningTenant(ctx, evalCtx, execCfg, oracle, localityFilter)
+	return dsp.setupAllNodesPlanningTenant(ctx, evalCtx, execCfg, oracle, localityFilter, nil /*txn*/)
+}
+
+// SetupAllNodesPlanningWithOracleAndFollowers is like
+// SetupAllNodesPlanningWithOracle but with a txn param that if non-nil allows
+// the flow to plan on followers.
+func (dsp *DistSQLPlanner) SetupAllNodesPlanningWithOracleAndFollowers(
```
Comment (miretskiy) [inline pkg/sql/distsql_plan_bulk.go:58]: would be nice to have tests for this... should be possible w/ injectable oracle implementations.  
TODO probably fine, but we really need to know for sure that this
stuff works correctly.

File: pkg/sql/distsql_plan_bulk.go
Comment (miretskiy) [inline pkg/sql/distsql_plan_bulk.go:1]: can you separate sql change into separate commit for ease of backporting?

File: pkg/ccl/backupccl/backup_job.go [hunk 3]
```diff
@@ -199,37 +206,59 @@ func backup(
...
+	var planCtx *sql.PlanningCtx
+
+	// We need to make a txn to setup planning since the txn is what will carry
+	// the read-as-of timestamp to the replica choice function, which needs it to
+	// determine if follower reads are possible. This txn is not intended to be
+	// used for anything -- no reads, no writes -- other than carrying its fixed
+	// timestamp from here to the replica choice oracle.
+	if err := execCtx.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		var planningTxn *kv.Txn
+		if backupUseFollowerReads.Get(execCtx.ExecCfg().SV()) {
```
Comment (miretskiy) [inline pkg/ccl/backupccl/backup_job.go:226]: does it make sense to have this if outside of db.Txn call to avoid starting txn if disabled?

**Review by miretskiy — COMMENTED** [review]
just some comments; haven't reviewed backup code.

File: pkg/ccl/backupccl/backup_job.go [hunk 4]
```diff
@@ -199,37 +206,59 @@ func backup(
...
+		// We don't return the compatible nodes here since PartitionSpans will
+		// filter out incompatible nodes.
+		planCtx, _, err = dsp.SetupAllNodesPlanningWithOracleAndFollowers(
+			ctx, evalCtx, execCtx.ExecCfg(), physicalplan.DefaultReplicaChooser, execLocality, planningTxn,
+		)
+		if err != nil {
+			return errors.Wrap(err, "failed to determine nodes on which to run")
+		}
+
+		backupSpecs, err = distBackupPlanSpecs(
```
Comment (stevendanna) [inline pkg/ccl/backupccl/backup_job.go:241]: It should be fine since we aren't _doing anything_ in the txn, but I'll note for visibility that in the case of encrypted backups this will be doing an external network call inside the txn.

File: pkg/ccl/backupccl/backup_job.go [hunk 5]
```diff
@@ -199,37 +206,59 @@ func backup(
...
+	// planCtx to dsp.Run below. However dsp.Run does not use the txn in planCtx,
+	// since it is passed its own txn (which will be nil), so the fact it is an
+	// already-committed txn is not a problem.
+	var planCtx *sql.PlanningCtx
+
+	// We need to make a txn to setup planning since the txn is what will carry
+	// the read-as-of timestamp to the replica choice function, which needs it to
+	// determine if follower reads are possible. This txn is not intended to be
+	// used for anything -- no reads, no writes -- other than carrying its fixed
+	// timestamp from here to the replica choice oracle.
```
Comment (stevendanna) [inline pkg/ccl/backupccl/backup_job.go:223]: I wonder if we want to format this in a way that might force future us to read it:

```
 //
 // DO NOT use this txn for anything -- no reads, no writes -- other than carrying its fixed
 // timestamp from here to the replica choice oracle.
```

File: pkg/ccl/backupccl/backup_job.go [hunk 6]
```diff
@@ -86,6 +87,12 @@ var BackupCheckpointInterval = settings.RegisterDurationSetting(
 	"the minimum time between writing progress checkpoints during a backup",
 	time.Minute)
 
+var backupUseFollowerReads = settings.RegisterBoolSetting(
+	settings.ApplicationLevel,
+	"bulkio.backup.plan_on_followers.enable",
```
Comment (stevendanna) [inline pkg/ccl/backupccl/backup_job.go:92]: The linter is gonna yell at you. s/enable/enabled/

File: pkg/ccl/backupccl/backup_job.go [hunk 7]
```diff
@@ -199,37 +206,59 @@ func backup(
...
+
+	// We need to make a txn to setup planning since the txn is what will carry
+	// the read-as-of timestamp to the replica choice function, which needs it to
+	// determine if follower reads are possible. This txn is not intended to be
+	// used for anything -- no reads, no writes -- other than carrying its fixed
+	// timestamp from here to the replica choice oracle.
+	if err := execCtx.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		var planningTxn *kv.Txn
+		if backupUseFollowerReads.Get(execCtx.ExecCfg().SV()) {
+			if err := txn.SetFixedTimestamp(ctx, backupManifest.EndTime); err != nil {
```
Comment (stevendanna) [inline pkg/ccl/backupccl/backup_job.go:227]: It appears from the test failure that we have a test that produces a backupManifest with an empty EndTime.  That shouldn't happen in production code AFAIK.  I suppose we could catch that case above this to produce a better error message but I don't feel too strongly about it.

**Review by stevendanna — APPROVED** [review]
I assume that this change is intended to be backportable. Given that, I think this solution is pretty reasonable.

I do wonder whether a different API from dsp would be better for this case where we want to create a plan now (when we may not be at a safe follower read ts yet) but plan do issue most of our requests in the future when we will be. But, any such change would not be backportable.

I feel compelled to ask for a test, but also a heavy multi-node test for a single piece of behaviour also hurts my soul a bit.

It appears yev and I were reviewing concurrently, so I tried to remove any comments that duplicated his.

File: pkg/sql/distsql_plan_bulk.go (hunk 1 above)
Comment (dt) [reply pkg/sql/distsql_plan_bulk.go:64]: Indeed, a timestamp is what I have and would prefer to pass, but we'd need to rip up several layers of API that are shared with all SQL queries to do that, which seems incompatible with the goals of low risk backports.

It is also worth noting that the txn contains a _mutable_ timestamp, that changes over the life of the txn, and the caller in the planning code is reading the _current_ value of it (with appropriate locking), so it probably is not as simple as s/kv.txn/hlc.timestamp/ but rather would need to be something more like s/kv.Txn/func() hlc.timestamp/ and then we'd need to consider if making this func means hot paths start allocating or something. Doable for sure but not the scope I want to introduce in a backportable change.

File: pkg/ccl/backupccl/backup_job.go (hunk 5 above)
Comment (dt) [reply pkg/ccl/backupccl/backup_job.go]: I changed my mind about this db.Txn and went with a more opinionated approach instead of making a mock txn backed by a sender that will fail all operations, loudly, if it is used.

File: pkg/ccl/backupccl/backup_job.go (hunk 3 above)
Comment (dt) [reply pkg/ccl/backupccl/backup_job.go]: I donno, seemed messy. I changed my mind again though and got rid of db.Txn entirely.

File: pkg/kv/mock_transactional_sender.go [hunk 8]
```diff
@@ -323,3 +325,27 @@ func (f MockTxnSenderFactory) LeafTransactionalSender(tis *roachpb.LeafTxnInputS
...
+// for reading its timestamp and little else.
+func MakeMockNoSenderTxnWithTimestamp(
+	ctx context.Context,
+	actx log.AmbientContext,
+	clock *hlc.Clock,
+	gatewayNodeID roachpb.NodeID,
+	dbctx DBContext,
+	ts hlc.Timestamp,
+) (*Txn, error) {
+	db := NewDBWithContext(actx, MakeMockTxnSenderFactory(func(
```
Comment (miretskiy) [inline pkg/kv/mock_transactional_sender.go:340]: I know (after reading the code) that `MakeMockTxnSenderFactory` are not much of a mock per say, and it just allows you to inject Send function that does whatever you want... I am concerned however, just how easy it is to make changes to a file called `mock_transaction_sender` and inadvertently introduce unwanted dependencies (such as testing, or testify or whatnot)

I think small code duplication to implement sender and txn factory might be safer/better.

File: pkg/sql/distsql_plan_bulk.go
Comment (dt) [reply pkg/sql/distsql_plan_bulk.go:1]: cdc doesn't actually need the sql change since cdc doesn't use `SetupAllNodesPlanning` but rather uses `NewPlanningCtxWithOracle` directly, which already has a txn param.

**Comment by dt** [comment]
I made the change I'd prefer to make -- switching the API on the planning to take a timestamp instead of a txn wrapping a timestamp -- over on https://github.com/cockroachdb/cockroach/pull/114574. I still think we merge and backport this change first, but I feel much better about this change knowing that that change compiles with the txn removed, since it proves we're not using this txn for anything else that we're not considering here.

File: pkg/kv/mock_transactional_sender.go (hunk 8 above)
Comment (dt) [reply pkg/kv/mock_transactional_sender.go]: I actually dropped mock change and just use a real txn from NewTxn now; I added the mock to prove we weren't mis-using the txn for anything other than just being a timestamp container, but now that I've proved that to myself via deleting the txn entirely in #114574 I no longer have any qualms about just doing NewTxn here for now.

File: pkg/ccl/backupccl/backup_job.go [hunk 9]
```diff
@@ -86,6 +86,12 @@ var BackupCheckpointInterval = settings.RegisterDurationSetting(
 	"the minimum time between writing progress checkpoints during a backup",
 	time.Minute)
 
+var backupUseFollowerReads = settings.RegisterBoolSetting(
```
Comment (benbardin) [inline pkg/ccl/backupccl/backup_job.go:89]: do we need this for merging to master? The only reason to disable it would be "hidden bug," right?

**Review by nvb — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114532)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @kvoli, @michae2, @miretskiy, and @stevendanna)

___
*[`pkg/ccl/backupccl/backup_job.go` line 93 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-uHGDk7PyaRheqSg:-NjO-uHGDk7PyaRheqSh:beds6at) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L93)):*
> ```Go
> 	"bulkio.backup.plan_on_followers.enabled",
> 	"assume utilization of follower reads when planning backup worker placement",
> 	true)
> ```

Should this default to true or false? Is the plan to switch this during the backport?
___
*[`pkg/ccl/backupccl/backup_job.go` line 209 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-57H5dTh7kUZL3qb:-NjO-57H5dTh7kUZL3qc:bmq6o8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L209)):*
> ```Go
> 
> 	var planningTS *kv.Txn
> 	if backupUseFollowerReads.Get(evalCtx.ExecCfg.SV()) {
> ```

Do we want to do this whenever the cluster setting is enabled, or only when paired with execution locality?
___
*[`pkg/ccl/backupccl/backup_job.go` line 210 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-GWx5GB3dcK89_6Z:-NjO-GWx5GB3dcK89_6_:bmhq5p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L210)):*
> ```Go
> 	var planningTS *kv.Txn
> 	if backupUseFollowerReads.Get(evalCtx.ExecCfg.SV()) {
> 		planningTS = kv.NewTxn(ctx, execCtx.ExecCfg().DB, roachpb.NodeID(dsp.GatewayID()))
> ```

Could we give this a comment explaining the purpose of the txn.
___
*[`pkg/ccl/backupccl/backup_job.go` line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-QaR6XWT8g_qjOlL:-NjO-QaR6XWT8g_qjOlM:b-u1m4bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L211)):*
> ```Go
> 	if backupUseFollowerReads.Get(evalCtx.ExecCfg.SV()) {
> 		planningTS = kv.NewTxn(ctx, execCtx.ExecCfg().DB, roachpb.NodeID(dsp.GatewayID()))
> 		if planningTS.SetFixedTimestamp(ctx, backupManifest.EndTime); err != nil {
> ```

Not for the backport, but it feels strange to me that `backupManifest.EndTime` would influence the planning of a backup. A backup is going to run for many minutes. Does it matter whether it starts backdated by 2s or 5s? What would happen if we just waited out the closed timestamp interval before starting the backup?
___
*[`pkg/sql/distsql_physical_planner.go` line 4735 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjNwlfV6Ycr7Qu2WLnR:-NjNwlfW3x75nZPYcZ8x:botpyo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9a106b5a672dabeaf16d59b657360b463fdb7a9/pkg/sql/distsql_physical_planner.go#L4735)):*
> ```Go
> 	evalCtx *extendedEvalContext,
> 	planner *planner,
> 	spanResolutionTS *kv.Txn,
> ```

It's strange to see a `kv.Txn` type called "...TS". I'm +1 on renaming this, but maybe we call it something else. `tsRequirementTxn`?

Alternatively, an interface like the following might go a long way in allowing you to avoid the fake transaction in `func backup`:
```go
package replicaoracle

type TimestampRequirement interface {  
  RequiredFrontier()  
}
```

You could add a second implementation that's just `type timestampRequirement hlc.Timestamp`.

But we don't need to do so in a backport if we want to keep this low risk.

NOTE: interface > method value (from #114574) because it avoids the heap allocation to close over the receiver for normal DistSQL planning.

NOTE: if you go with this, watch out for packing a nil `*kv.Txn` into the interface and having a non-nil interface wrap a nil txn.
___
*[`pkg/sql/distsql_plan_bulk.go` line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjNzkHV6jHYXMUz3Y2f:-NjNzkHV6jHYXMUz3Y2g:bbwzc1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9a106b5a672dabeaf16d59b657360b463fdb7a9/pkg/sql/distsql_plan_bulk.go#L49)):*
> ```Go
> 	localityFilter roachpb.Locality,
> ) (*PlanningCtx, []base.SQLInstanceID, error) {
> 	if dsp.codec.ForSystemTenant() {
> ```

Should this just call `SetupAllNodesPlanningWithOracleAndFollowers`?
___
*[`pkg/sql/distsql_plan_bulk.go` line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjNzuz_Eoth7wWAOflS:-NjNzuz_Eoth7wWAOflT:bjqlz9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9a106b5a672dabeaf16d59b657360b463fdb7a9/pkg/sql/distsql_plan_bulk.go#L120)):*
> ```Go
> 	oracle replicaoracle.Oracle,
> 	localityFilter roachpb.Locality,
> 	txn *kv.Txn,
> ```

Did you mean to do the same rename here?

File: pkg/sql/distsql_plan_bulk.go (hunk 1 above)
Comment (benbardin) [reply pkg/sql/distsql_plan_bulk.go]: Could we not fix master as correctly as possible, and then think about a backport separately?

File: pkg/ccl/backupccl/backup_job.go (hunk 9 above)
Comment (dt) [reply pkg/ccl/backupccl/backup_job.go:89]: Yes, we do. We like having a toggle for any change both in case there is an unforeseen bug like you mention, but also just so we can rule such a thing out as well: if we see something behave unexpectedly in the next several months, we might wonder _if_ it is a hidden bug, or even perhaps not a bug but just a new way it behaves, and being able to toggle it off quickly is a way to cheaply _rule out_ that possibility.

Comment (benbardin) [reply pkg/ccl/backupccl/backup_job.go:89]: 👍

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114532)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @kvoli, @michae2, @miretskiy, @nvanbenschoten, and @stevendanna)

___
*[`pkg/ccl/backupccl/backup_job.go` line 93 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-uHGDk7PyaRheqSg:-NjODPS-38du2frAkvFN:bds4pth) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L93)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this default to true or false? Is the plan to switch this during the backport?
</blockquote></details>

Yeah, intend to have it be true on master and 23.2 backport, then flip to false in 23.1 backport.
___
*[`pkg/ccl/backupccl/backup_job.go` line 209 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-57H5dTh7kUZL3qb:-NjODWWm5yv-IIy9gLsR:brbswt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L209)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we want to do this whenever the cluster setting is enabled, or only when paired with execution locality?
</blockquote></details>

I think for all backup planning? 

I think ideally we either consider followers or not in whatever set of eligible nodes we'll plan on, regardless of how we got that set / if it is the set of all nodes or just those in some locality. I'd rather not have completely different planning behaviors to think about if the set is filtered or not if we can avoid it.

Obviously the fact that the set of all nodes always includes all leaseholders means it will be different 
___
*[`pkg/ccl/backupccl/backup_job.go` line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-QaR6XWT8g_qjOlL:-NjOO4AA6FZvg2NPGdsO:b4t17ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L211)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Not for the backport, but it feels strange to me that `backupManifest.EndTime` would influence the planning of a backup. A backup is going to run for many minutes. Does it matter whether it starts backdated by 2s or 5s? What would happen if we just waited out the closed timestamp interval before starting the backup?
</blockquote></details>

`EndTime` is the hlc timestamp as of which we will read when we send export requests, and is set by the `AS OF SYSTEM TIME` in the backup statement. It is called "end" time to distinguish it from _start_ time when we are reading deltas during an incremental backup. 

I think this is the correct timestamp to influence planning? We need replicas that can serve requests up to and including `EndTime`, so that's the timestamp we'd provide when deciding it we can use some followers or not.
___
*[`pkg/sql/distsql_physical_planner.go` line 4735 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjNwlfV6Ycr7Qu2WLnR:-NjOSJKF44RcoBXudc8C:b9kbmfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9a106b5a672dabeaf16d59b657360b463fdb7a9/pkg/sql/distsql_physical_planner.go#L4735)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's strange to see a `kv.Txn` type called "...TS". I'm +1 on renaming this, but maybe we call it something else. `tsRequirementTxn`?

Alternatively, an interface like the following might go a long way in allowing you to avoid the fake transaction in `func backup`:
```go
package replicaoracle

type TimestampRequirement interface {  
  RequiredFrontier()  
}
```

You could add a second implementation that's just `type timestampRequirement hlc.Timestamp`.

But we don't need to do so in a backport if we want to keep this low risk.

NOTE: interface > method value (from #114574) because it avoids the heap allocation to close over the receiver for normal DistSQL planning.

NOTE: if you go with this, watch out for packing a nil `*kv.Txn` into the interface and having a non-nil interface wrap a nil txn.
</blockquote></details>

I'd like to keep this PR pretty narrowly focused and thus don't want to change the types in this PR. I will copy the interface suggestion over to #114574 instead. 

As far for the name, I agree it is strange but I kinda _like_ the strange here? This is a strange txn, that isn't supposed to be used as a txn, so having a strange name that doesn't make it sounds like it is a txn seems nice to me.
___
*[`pkg/sql/distsql_plan_bulk.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114532#gh-1395017023:-NjOeLNcAR9S9udaJmBA:bi323x4) ([raw file](https://github.com/cockroachdb/cockroach/blob/372f95f5e14945fc540d4d59051f892b8cf088bc/pkg/sql/distsql_plan_bulk.go#L64)):*
<details><summary><i>Previously, benbardin (Ben Bardin) wrote…</i></summary><blockquote>

Could we not fix master as correctly as possible, and then think about a backport separately?
</blockquote></details>

I think we are doing that, just in stages. We don't like making bespoke changes for release branches that never ran on master when we can avoid it; those are generally believed to be higher risk. When possible, we like to make the changes on master, broken up into pieces that each make the situation better, with the boundaries between these being such that we can backport some prefix of them.

We're talking about where we think master should end up _after_ this change, but this change is still making master better than it is now. If we were making master _worse_ in the name of a a backport I'd be concerned, but we aren't. We think the change we're making here is making master more correct, and then there is _another_ change we could make that would make it _even more_ correct (but would not backport), but I don't think the fact that we have that second change should prevent this change.
___
*[`pkg/sql/distsql_plan_bulk.go` line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjNzkHV6jHYXMUz3Y2f:-NjO_UMHEV6f1NMmjqH4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9a106b5a672dabeaf16d59b657360b463fdb7a9/pkg/sql/distsql_plan_bulk.go#L49)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this just call `SetupAllNodesPlanningWithOracleAndFollowers`?
</blockquote></details>

Done.

**Review by nvb — COMMENTED** [review]
> Reviewed 2 of 2 files at r6, 1 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114532)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @kvoli, @michae2, @miretskiy, and @stevendanna)

___
*[`pkg/ccl/backupccl/backup_job.go` line 209 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-57H5dTh7kUZL3qb:-NjStMAT-Ego-7rp0feh:b-z5rb9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L209)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think for all backup planning? 

I think ideally we either consider followers or not in whatever set of eligible nodes we'll plan on, regardless of how we got that set / if it is the set of all nodes or just those in some locality. I'd rather not have completely different planning behaviors to think about if the set is filtered or not if we can avoid it.

Obviously the fact that the set of all nodes always includes all leaseholders means it will be different
</blockquote></details>

Won't this change the planning of normal backup to bunch processors closer to the gateway, causing imbalance? We talked earlier about how DistSQL doesn't do its own load-balancing, it just relies on the balance of leaseholders to spread load across the cluster. We intentionally want to ignore this placement when execution locality is in use, because we favor local execution (where we typically expect to only have a single replica in the entire locality) over maximal balance.

By the way, the case I'm mostly concerned about this this isn't multi-region — it's a single region cluster with three availability zones. This is the most common configuration of CRDB (at least in cloud, citation needed elsewhere) and it results in all three zones having a replica of every range. I'd expect that with this change as it currently stands, we would only ever plan backup processors in the zone with the gateway that executed the BACKUP statement. That would triple the load in that zone and remove it from the other two. We don't want this change to have such an effect, do we?

EDIT: given the discussion below, I guess this only affects `BACKUP` with `AS OF SYSTEM TIME`, not every backup. That's still a larger set of backups then I think we're intending to change the behavior of.
___
*[`pkg/ccl/backupccl/backup_job.go` line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjO-QaR6XWT8g_qjOlL:-NjSv65k9y5lUG3_u5fT:b-2zbfzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/922705586c42cbef35b594c3d1169e9290651ff6/pkg/ccl/backupccl/backup_job.go#L211)):*
> I think this is the correct timestamp to influence planning? We need replicas that can serve requests up to and including `EndTime`, so that's the timestamp we'd provide when deciding it we can use some followers or not.

Agreed that local replicas may not be immediately able to serve Export requests at this time if `AS OF SYSTEM TIME` is not used. However, they will be able to in 3-4 seconds. So my question is whether we like the fact that this 3-4 second period is dictating where a potentially long-running backup process is planned. If we expect all backups to take minutes, it might be faster to just wait out the 3-4 seconds at the beginning and then use a more optimal plan.

Just a thought, and one that is mostly separate from the rest of this PR.
___
*[`pkg/sql/distsql_physical_planner.go` line 4735 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjNwlfV6Ycr7Qu2WLnR:-NjSvkiJ1DowA4Wd4kym:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9a106b5a672dabeaf16d59b657360b463fdb7a9/pkg/sql/distsql_physical_planner.go#L4735)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'd like to keep this PR pretty narrowly focused and thus don't want to change the types in this PR. I will copy the interface suggestion over to #114574 instead. 

As far for the name, I agree it is strange but I kinda _like_ the strange here? This is a strange txn, that isn't supposed to be used as a txn, so having a strange name that doesn't make it sounds like it is a txn seems nice to me.
</blockquote></details>

:thumbsup: 
___
*[`pkg/sql/distsql_plan_bulk.go` line 56 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114532#-NjSp4iM5sA8aZelTN1c:-NjSp4iM5sA8aZelTN1d:b9s221c) ([raw file](https://github.com/cockroachdb/cockroach/blob/27839e8d3d5081b478836aa9e6e00267026a024e/pkg/sql/distsql_plan_bulk.go#L56)):*
> ```Go
> // which can be used to pass a txn which, if non-nil, is used to convey the
> // timestamp at which the processors in the flow will read for the purposes of
> // determining if
> ```

This comment cuts off.

**Comment by dt** [comment]
After considering the potential effects @nvanbenschoten highlighted in some cases w.r.t non-locality-filtered planning if we just drop leaseholder preferences, I think I might prefer the approach in https://github.com/cockroachdb/cockroach/pull/114710 instead (it also avoids this sketching non-txn txn, _and_ solves the work distribution questions, including for non-balanced leaseholder configs).

**Comment by dt** [comment]
Concretely, I propose closing this PR in favor of the approach in https://github.com/cockroachdb/cockroach/pull/114710

**Comment by dt** [comment]
Superseded by #114710
